File: 81. Search in Rotated Sorted Array II/solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        return self.mysearch(nums, target, 0, len(nums)-1)
        
    def mysearch(self, nums, target, low, high):
        # base case
        if high - low <= 1:
            return target in nums[low:high+1]
        
        # check mid
        mid = (low+high)//2
        if nums[mid] == target:
            return True
        
        # check if there is repetition -> tends to O(n) solution
        if nums[low] == nums[high] or nums[mid] == nums[high]:
            return self.mysearch(nums, target, low, mid-1) or self.mysearch(nums, target, mid+1, high)
        
        # look for the sorted side and choose the correct half
        elif nums[mid] < nums[high]:
            if nums[mid] < target <= nums[high]:
                return self.mysearch(nums, target, mid+1, high)
            else:
                return self.mysearch(nums, target, low, mid-1)
        else:
            if nums[low] <= target < nums[mid]:
                return self.mysearch(nums, target, low, mid-1)
            else:
                return self.mysearch(nums, target, mid+1, high)
                
        

File: 81. Search in Rotated Sorted Array II/solution.java
class Solution {
    public boolean search(int[] nums, int target) {
        if (nums.length == 0) {
            return false;
        }
        return binarySearch(nums, target, 0, nums.length-1);
    }
    
    public boolean binarySearch(int[] nums, int target, int low, int high) {
        // base condition
        if (high - low == 2) {
            if (nums[high] == target) {
                return true;
            } else if (nums[low] == target) {
                return true;
            } else if (nums[low + 1] == target) {
                return true;
            } else {
                return false;
            }
        } else if (high - low == 1) {
            if (nums[high] == target) {
                return true;
            } else if (nums[low] == target) {
                return true;
            } else {
                return false;
            }
        } else if (high == low) {
            if (nums[high] == target) {
                return true;
            } else {
                return false;
            }
        }
        
        // divide and conquer
        int mid = (high + low)/2;
        if (nums[mid] == target) {
            return true;
        } 
        
        // the comparison below relies on the fact that it can decisively find 
        // the unrotated half in the rotated array by comparing low, mid and high
        // if elements repeat among low, mid or high, this cannot be found clearly
        // In such a case, linear search on both halves 
        else if (nums[low] == nums[high] || nums[low] == nums[mid] || nums[high] == nums[mid]) {
            return binarySearch(nums, target, low, mid-1) ||
                binarySearch(nums, target, mid+1, high);
            
        } else if (nums[mid] > nums[low]) {
            if (target < nums[mid] && target >= nums[low]) {
                return binarySearch(nums, target, low, mid-1);
            } else {
                return binarySearch(nums, target, mid+1, high);
            }
        } else {
            if (target <= nums[high] && target > nums[mid]) {
                return binarySearch(nums, target, mid+1, high);
            } else {
                return binarySearch(nums, target, low, mid-1);
            }
        }
    }
}
